From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Andrew Kaster <andrew@ladybird.org>
Date: Mon, 26 Aug 2024 21:35:45 -0600
Subject: [PATCH] [CMake] Allow ExternalProjects to use console with Ninja
 generator

Also stop forcing the SwiftMacros project to build every rebuild.
---
 Sources/CMakeLists.txt               | 13 ++++++++++++-
 Sources/TestingMacros/CMakeLists.txt | 14 +++++++++++++-
 2 files changed, 25 insertions(+), 2 deletions(-)

diff --git a/Sources/CMakeLists.txt b/Sources/CMakeLists.txt
index 83a9a2b2330c9847ebac89e273c6ade9bf398a25..d917509ba47c2ddaf6b7a2d7d17de2afaabc68fd 100644
--- a/Sources/CMakeLists.txt
+++ b/Sources/CMakeLists.txt
@@ -43,10 +43,21 @@ if(SwiftTesting_MACRO STREQUAL "<auto>")
   # Build and install the plugin into the current build directry.
   set(SwiftTesting_MACRO_INSTALL_PREFIX "${CMAKE_BINARY_DIR}")
 
+  # Allow the Ninja generators to output messages as they happen by assigning
+  # these jobs to the 'console' job pool
+  set(SwiftTesting_ConsoleAccess "")
+  if(CMAKE_GENERATOR MATCHES "^Ninja")
+    set(SwiftTesting_ConsoleAccess
+      USES_TERMINAL_CONFIGURE YES
+      USES_TERMINAL_BUILD YES
+      USES_TERMINAL_INSTALL YES)
+  endif()
+
   ExternalProject_Add(TestingMacros
     PREFIX "tm"
     SOURCE_DIR "${CMAKE_CURRENT_SOURCE_DIR}/TestingMacros"
-    BUILD_ALWAYS ON
+    BUILD_ALWAYS OFF
+    ${SwiftTesting_ConsoleAccess}
     CMAKE_ARGS
       -DCMAKE_MAKE_PROGRAM=${SwiftTesting_MACRO_MAKE_PROGRAM}
       -DCMAKE_Swift_COMPILER=${SwiftTesting_MACRO_Swift_COMPILER}
diff --git a/Sources/TestingMacros/CMakeLists.txt b/Sources/TestingMacros/CMakeLists.txt
index a45663025a43adf39d17f0d94d692c9bf6092194..aa9939079ff18dee00288bf7fb0c92c124438228 100644
--- a/Sources/TestingMacros/CMakeLists.txt
+++ b/Sources/TestingMacros/CMakeLists.txt
@@ -28,11 +28,23 @@ if(SwiftTesting_BuildMacrosAsExecutables)
   # When building the macro plugin as an executable, clone and build
   # swift-syntax.
   include(FetchContent)
+
+  # Allow the Ninja generators to output messages as they happen by assigning
+  # these jobs to the 'console' job pool
+  set(SwiftTesting_ConsoleAccess "")
+  if(CMAKE_GENERATOR MATCHES "^Ninja")
+    set(SwiftTesting_ConsoleAccess
+      USES_TERMINAL_CONFIGURE YES
+      USES_TERMINAL_BUILD YES
+      USES_TERMINAL_INSTALL YES)
+  endif()
+
   set(FETCHCONTENT_BASE_DIR ${CMAKE_BINARY_DIR}/_d)
   # TODO: Update GIT_TAG to the 6.0 release tag once it is available.
   FetchContent_Declare(SwiftSyntax
     GIT_REPOSITORY https://github.com/swiftlang/swift-syntax
-    GIT_TAG 27b74edd5de625d0e399869a5af08f1501af8837)
+    GIT_TAG 27b74edd5de625d0e399869a5af08f1501af8837
+    ${SwiftTesting_ConsoleAccess})
   FetchContent_MakeAvailable(SwiftSyntax)
 endif()
 
